
---
output:
 md_document:
  variant: gfm
  toc: true
---


# Tutorial for analyzing GEO bulk RNA-Seq datasets with Pasta  

**Author:** Jérôme Salignon  
**Date:** `r Sys.Date()`  


# Introduction

This document provides an example analysis using Pasta on dataset **GSE149694** from Liu & Polo, *Nature*, 2020. The study, titled _"Reprogramming roadmap reveals route to human induced trophoblast stem cells"_, is available online at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149694), [Nature](https://www.nature.com/articles/s41586-020-2734-6), and on the [HRPI website](http://hrpi.ddnetbio.com/).

# Data acquisition, age-prediction, and processing

In this section, we load libraries, download the GEO ExpressionSet, obtain phenotype data, and predict age scores. 

## Loading libraries

```{r setup, message=FALSE, warning=FALSE}
library(pasta)
library(jsutil)
library(magrittr)
library(data.table)
library(ggplot2)
library(gtools)

# Create output directory if needed
if (!dir.exists("../output")) {
  dir.create("../output")
}
```
You can install any missing packages using `install.packages()` for CRAN packages (`magrittr`, `data.table`, `ggplot2`, `gtools`), or `devtools::install_github()` for GitHub packages (`pasta`, `jsutil`).


## Getting GEO ExpressionSet and predict age scores

```{r get-ES, message=FALSE, warning=FALSE}

file_LiuPolo2020 = '../output/ES_LiuPolo2020.rds'
if(!file.exists(file_LiuPolo2020)){
  ES <- getting_GEO_ES_for_age_model('GSE149694') %T>% pdim  # Dimensions: 8113 genes x 32 samples
  saveRDS(ES, file_LiuPolo2020)
}

ES = readRDS(file_LiuPolo2020)
pdata <- getting_pdata_with_age_scores(ES)
```

## Reshaping to long format and cleaning the metadata

Here we reshape the data to long format and extract information such as condition, time, and replicate details from the title.

```{r reshape-data, message=FALSE, warning=FALSE}
pdata1 <- pdata[, c('title', 'PASTA', 'REG', 'CT46')]
dt <- melt(pdata1, id.vars = 'title', variable.name = 'model_type', value.name = 'age_score')
dt[, title := gsub('RNA-seq_', '', title)]
dt[, condition := gsub('-.*', '', title)]
dt[, time := gsub('.*-(.*)_rep.*', '\\1', title)]
dt[, time := factor(time, levels = gtools::mixedsort(unique(dt$time)))]
dt[, rep := gsub('.*rep', '', title)]
dt[, title := NULL]
dt[, time1 := as.integer(time)]
```


## Adding initial time points

The initial time points (Fibroblast, day 3 and 7) are the same between all conditions ('NHSM', '5iLAF', 'Primed', 't2iLGoY', 'RSeT'). We add these time point to each condition so we can then compute correlation.

```{r load-adjust-data, message=FALSE, warning=FALSE}
dt_pasta <- dt[model_type == 'PASTA']

# Separate fibroblast and non-fibroblast data
dt_fibro <- dt_pasta[condition == 'Fibroblast']
dt_liuPolo <- dt_pasta[condition != 'Fibroblast']

# For each non-fibroblast condition, append a copy of the fibroblast data with the condition overwritten
for(condition1 in unique(dt_liuPolo$condition)){
  dt_fibro1 <- copy(dt_fibro)[, condition := condition1]
  dt_liuPolo <- rbind(dt_liuPolo, dt_fibro1)
}

# Reformat time variables
dt_liuPolo[, time1 := time %>% gsub('D', '', .) %>% gsub('P3', '30', .) %>% 
  gsub('P10', '30', .) %>% as.integer]
# Display unique condition and time mappings
unique(dt_liuPolo[, c('condition', 'time', 'time1')])[order(condition, time1)]
```


# Results

## Computing correlations

We calculate both the Pearson and Spearman correlations coefficients between predicted age scores and the true time values.

```{r compute-correlations, message=FALSE, warning=FALSE}
dt_cor <- dt_liuPolo[, .(
  PCC = cor(age_score, time1, method = 'pearson'), 
  SCC = cor(age_score, time1, method = 'spearman')
), by = c('model_type', 'condition')][order(SCC)]
print(dt_cor)
```

We can also display overall correlation values.

```{r overall-correlation, message=FALSE, warning=FALSE}
cat("Pearson correlation between time1 and age_score:",
    cor(dt_liuPolo$time1, dt_liuPolo$age_score), "\n")
cat("Spearman correlation between time1 and age_score:",
    cor(dt_liuPolo$time1, dt_liuPolo$age_score, method = 'spearman'), "\n")
```


## Visualization

Finally, we visualize the timecourse of the age-effect predictions.

```{r plot-timecourse, message=FALSE, warning=FALSE}
# Plot predicted age scores over time for the Liu & Polo dataset
p_LiuPolo2020 <- ggplot(dt_liuPolo, aes(x = time, y = age_score)) + 
  geom_point(size = 1, alpha = 0.7) +
  ggtitle('Reprogramming Timecourse') +
  theme_minimal()  # Replace with your custom theme if available

print(p_LiuPolo2020)
```


