
---
title: "Pasta Analysis Tutorial: Human Retina Horizontal Cells"
author: "Jérôme Salignon"
date: "`r Sys.Date()`"
output:
  md_document:
    variant: gfm
    toc: true
---

# Introduction

# Introduction

This tutorial demonstrates an analysis using the **Pasta** package on the dataset **GSE135133** from Orozco, *Cell Report*, 2020.  
The study, *"Integration of eQTL and a Single-Cell Atlas in the Human Eye Identifies Causal Genes for Age-Related Macular Degeneration"*, is available at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE41714), [CellxGene](https://cellxgene.cziscience.com/collections/939769a8-d8d2-4d01-abfc-55699893fd49) and [Cell Report](https://www.cell.com/cell-reports/fulltext/S2211-1247%2819%2931747-4). In particular, the dataset "Horizontal cells of human eye" of 1,875 cells was downloaded from CellxGene. 


# Data Acquisition, Age-Prediction, and Processing

In this section, we load the example Seurat object, process the metadata, filter cell types, and create pseudobulk samples for age prediction.

## Loading Libraries and Dataset

```{r setup, include=FALSE}
library(magrittr)
library(jsutil)
library(pasta)
```

```{r load-data, message=FALSE, warning=FALSE}
# Load example dataset of retina horizontal cells
data(seu_orozco_2020_retina_horizontal_cells)
seu <- seu_orozco_2020_retina_horizontal_cells %T>% pdim  # Dimensions: 57,596 genes x 1,875 cells
rm(seu_orozco_2020_retina_horizontal_cells)
```

## Processing the Metadata

We adjust the metadata to extract age information and cell type annotations.

```{r process-metadata, message=FALSE, warning=FALSE}
# Process the metadata: remove extra characters from development_stage and convert cell type to character.
seu$age  <- seu$development_stage %>% gsub("-year.*", "", .) %>% gsub("-", " ", .)
seu$type <- seu$cell_type %>% as.character
```

Distribution of ages
```{r process-metadata, message=FALSE, warning=FALSE}
seu@meta.data$age %>% table
```

Distribution of cell types
```{r process-metadata, message=FALSE, warning=FALSE}
seu@meta.data$cell_type %>% table
```

Preview cell type filtering (dry-run)
```{r process-metadata, message=FALSE, warning=FALSE}
seu %>% filter_cell_types_in_seu_object(n_cell_min = 500, dry_run = TRUE, verbose = TRUE)
```


## Filtering and Creating Pseudobulk Samples

We filter the Seurat object to retain only cell types with at least 500 cells, and then create pseudobulk samples for age prediction.

```{r filter-pseudobulk, message=FALSE, warning=FALSE}
# Filter the Seurat object to keep only cell types with at least 500 cells
seu %<>% filter_cell_types_in_seu_object %T>% pdim

# Create pseudobulk samples using a chunk size of 1000 cells
dt_age_pred <- making_pseudobulks_and_predict_age(seu, chunk_size = 1000)
print(dt_age_pred)

# Predict age using multiple pseudobulk chunk sizes
v_chunk_sizes <- 2^(0:10)
dt_age_pred <- predicting_age_multiple_chunks(seu, v_chunk_sizes)
```

dt_age_pred <- predicting_age_multiple_chunks(seu, v_chunk_sizes)

dt_age_pred[, .N, chunk_size]


# Results

In this section, we compute correlations between the true age and predicted age scores and visualize the results.

## Correlation Analysis

```{r compute-correlations, message=FALSE, warning=FALSE}
# Compute correlations by chunk size for different modeling strategies
cur_dt_cor <- dt_age_pred[, .(REG = cor(age, REG),
                               PASTA = cor(age, PASTA),
                               TC46 = cor(age, CT46)), by = chunk_size]

# Reshape the data into long format for plotting
cur_dt1 <- melt(cur_dt_cor, id.vars = 'chunk_size', variable.name = 'Modeling_strategy', value.name = 'PCC')

# Optionally, add or adjust factor levels for the modeling strategies
model_levels <- c('REG', 'TC46', 'PASTA')
cur_dt1$Modeling_strategy <- factor(cur_dt1$Modeling_strategy, levels = model_levels)
print(cur_dt1)
```

## Visualization

We now plot the Pearson correlation coefficients (PCC) for the different modeling strategies across chunk sizes.

```{r plot-correlations, message=FALSE, warning=FALSE}
# Plot PCC vs. log2(chunk_size) for different modeling strategies
p1 <- ggplot(cur_dt1, aes(x = log2(chunk_size), y = PCC, colour = Modeling_strategy)) +
  geom_point(size = 2) +
  geom_line() +
  scale_colour_manual(values = c('PASTA' = 'red2', 'REG' = 'dodgerblue', 'TC46' = 'forestgreen')) +
  ggtitle("Correlation between True Age and Predicted Age Scores") +
  xlab("log2(Chunk Size)") +
  ylab("Pearson Correlation Coefficient (PCC)") +
  theme_minimal()

print(p1)
```
