---
output:
  md_document:
    variant: gfm
    toc: true
---

# Tutorial for analyzing mouse microarray datasets with Pasta

**Author:** Jérôme Salignon
**Date:** `r Sys.Date()`

# Introduction

This tutorial demonstrates an analysis using the **Pasta** package on the mouse dataset **GSE40156** from Yin, *Methods Mol Biol*, 2015.  
The study, *"Generation of Aorta Transcript Atlases of Wild-Type and Apolipoprotein E-null Mice by Laser Capture Microdissection-Based mRNA Expression Microarrays"*, is available at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40156) and [Methods in Mouse Atherosclerosis](https://doi.org/10.1007/978-1-4939-2929-0_20).


# Data acquisition, processing, and age-prediction 

In this section, we load libraries, download the GEO ExpressionSet, process the count matrix and metadata, and predict age scores.

## Loading libraries

```{r setup, message=FALSE, warning=FALSE}
library(pasta)
library(jsutil)
library(magrittr)
library(data.table)
library(ggplot2)

library(GEOquery)
library(AnnotationDbi)

library(org.Mm.eg.db)
library(mouse4302.db)

# Create output directory if needed
if (!dir.exists("../output")) {
  dir.create("../output")
}
```

You can install any missing packages using install.packages() for CRAN packages (magrittr, data.table, ggplot2, gtools), or devtools::install_github() for GitHub packages (pasta, jsutil) or BiocManager::install for Bioconductor packages (org.Mm.eg.db, mouse4302.db).


## Downloading GSE40156 from GEO

```{r download-gse40156, message=FALSE, warning=FALSE}

file_Yin2015_mouse = '../output/ES_Yin2015_mouse.rds'
if(!file.exists(file_Yin2015_mouse)){

  # Downloading a list of ExpressionSets from GEO
  lES <- getGEO('GSE40156', GSEMatrix = TRUE, getGPL = FALSE)

  # Keeping the Affymetrix Mouse Genome 430 2.0 Array (GPL1261) data
  ES <- lES[["GSE40156-GPL1261_series_matrix.txt.gz"]]

  saveRDS(ES, file_Yin2015_mouse)
} else {
  ES <- readRDS(file_Yin2015_mouse)
}

mat <- exprs(ES) %T>% pdim  # 45,101 probes, 42 samples

```


## Converting gene ids using AnnotationDbi

In this section, we use **AnnotationDbi** and the **mouse4302.db** package to convert probe identifiers from the Mouse430 2 array to Ensembl gene identifiers.

```{r convert-probe-to-ensembl, message=FALSE, warning=FALSE}

file_dt_ids_affy_mouse430 <- "../output/dt_ids__affy_mouse430_2.rds"
if (!file.exists(file_dt_ids_affy_mouse430)) {

  dt_ids <- AnnotationDbi::select(
    mouse4302.db,
    keys    = rownames(mat),
    keytype = "PROBEID",
    columns = c("ENSEMBL")
  ) %>% setDT %T>% pnrow

  colnames(dt_ids) <- c("affy_mouse430_2", "ensembl_gene_id")

  # Keep only probes with an Ensembl identifier
  dt_ids <- dt_ids[!is.na(ensembl_gene_id), ] %T>% pnrow

  # Save conversion table for future use
  saveRDS(dt_ids, file = file_dt_ids_affy_mouse430)

} else {
  dt_ids <- readRDS(file_dt_ids_affy_mouse430)
}

# Rename rows of the matrix to Ensembl identifiers using helper from jsutil / pasta
mat <- renaming_rows_in_mat_after_gene_id_conversion(mat, dt_ids, 
    species = "mouse") %T>% pdim   # 18,353 genes, 42 samples
```


## Preparing data for age-prediction

Here we filter the matrix for one-to-one orthologs between mouse and human and keep the human name, then we preprocess the matrix for age-prediction by imputing missing genes with median and doing rank normalization. 

```{r preprocess-matrix-for-age-prediction, message=FALSE, warning=FALSE}

mat %<>% keeping_mouse_one2one_orthologs %T>% pdim # 3,010 genes, 42 samples

mat %<>% filtering_age_model_genes_and_rank_norm %T>% pdim # 8,113 genes, 42 samples

```

## Creating metadata and predict ages

```{r extract-pdata, message=FALSE, warning=FALSE}

# Phenotype data from GEOquery ExpressionSet
pdata <- pData(ES) %>% setDT %>% copy
pdata %<>% .[, c("geo_accession", "age:ch1", "tissue:ch1", "strain:ch1")] 
colnames(pdata) <- c("geo_accession", "age", "tissue", "strain")

# Sanity check
all.equal(colnames(mat), pdata$geo_accession) # TRUE

# Adding Pasta age predictions
pdata %<>% adding_age_preds_to_pdata(t(mat)) %T>% pdim 

# Keeping only samples with non missing age
pdata <- pdata[!is.na(age)] %T>% pdim

# Age is counted in weeks; we convert it to integer
pdata[, age := gsub(" weeks", "", age) %>% as.integer]

```


# Results

Here we compute and visualize the Pearson correlation coefficients between chronological age and Pasta age scores, stratified by tissue and strain.

## Computing correlations

```{r correlations-by-group, message=FALSE, warning=FALSE}
dt_cor = pdata[, .(PCC = cor(age, Pasta)), c('tissue', 'strain')][order(-PCC)]
dt_cor[, strain := gsub('.*Apoe.*', 'Apoe', strain)]
dt_cor[, strain := gsub('.*C57B.*', 'WT', strain)]
dt_cor[, condition := paste0(tissue, ', ', strain)]
dt_cor[, condition := factor(condition, levels = dt_cor$condition %>% rev)]
print(dt_cor)
```


## Visualization

```{r plot-pcc, message=FALSE, warning=FALSE}
p1 <- ggplot(dt_cor, aes(x = PCC, y = condition)) +
  geom_point(size = 5) +
  theme_minimal(base_size = 18) +
  ggtitle("Mouse microarray (GSE40156)") +
  ylab("Tissue, strain") +
  xlab("Pearson correlation coefficient")

print(p1)
```

