---
output:
  md_document:
    variant: gfm
    toc: true
---

# Tutorial for analyzing mouse microarray datasets with Pasta

**Author:** Jérôme Salignon
**Date:** `r Sys.Date()`

# Introduction

This tutorial demonstrates an analysis using the **Pasta** package on the mouse dataset **GSE40156** from Hu, *Immunity*, 2015.  
The study, *"Implications of time-series gene expression profiles of replicative senescence"*, is available at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE41714) and [Aging Cell](https://onlinelibrary.wiley.com/doi/10.1111/acel.12087).


We will

1. Download the series matrix for GSE40156
2. Convert Affymetrix Mouse430 2 probe identifiers to Ensembl gene identifiers
3. Restrict the matrix to one to one mouse orthologs used in the Pasta age model
4. Predict age with Pasta and compute correlations between true age and predicted age
5. Summarize performance per tissue and per strain and visualize the results

# Data acquisition and preprocessing

## Loading libraries

```r
```{r setup, message=FALSE, warning=FALSE}
library(pasta)
library(jsutil)
library(magrittr)
library(data.table)
library(ggplot2)

library(GEOquery)
library(AnnotationDbi)

# The following Bioconductor packages are required for mouse430 2 annotation
# BiocManager::install("org.Mm.eg.db")
# BiocManager::install("mouse4302.db")
library(org.Mm.eg.db)
library(mouse4302.db)

# Create output directory if needed
if (!dir.exists("../output")) {
  dir.create("../output")
}

# Directory for validation objects
out_dir_2_PASTA_valid <- "../output/"
```

You can install missing CRAN packages with `install.packages()`
and Bioconductor packages with `BiocManager::install()` as commented above.

## Downloading GSE40156 from GEO

```r
```{r download-gse40156, message=FALSE, warning=FALSE}
cur_gse <- "GSE40156"

# Download series matrix without platform annotation tables
lES <- getGEO(cur_gse, GSEMatrix = TRUE, getGPL = FALSE)

# GSE40156 is on the Affymetrix Mouse Genome 430 2.0 Array (GPL1261)
names(lES)
ES <- lES[["GSE40156-GPL1261_series_matrix.txt.gz"]]

# Expression matrix, probes x samples
mat <- exprs(ES) %T>% pdim  # about 45,101 probes, 42 samples
```

## Converting probe identifiers to Ensembl gene identifiers

We convert probe identifiers from the Mouse430 2 array to Ensembl gene identifiers.
To make future runs faster we save the conversion table as an RDS file and reuse it when available.

```r
```{r convert-probe-to-ensembl, message=FALSE, warning=FALSE}
file_dt_ids_affy_mouse430 <- "../output/dt_ids__affy_mouse430_2.rds"

if (!file.exists(file_dt_ids_affy_mouse430)) {

  dt_ids <- AnnotationDbi::select(
    mouse4302.db,
    keys    = rownames(mat),
    keytype = "PROBEID",
    columns = c("ENSEMBL")
  ) %>%
    setDT %T>% pnrow

  colnames(dt_ids) <- c("affy_mouse430_2", "ensembl_gene_id")

  # Keep only probes with an Ensembl identifier
  dt_ids <- dt_ids[!is.na(ensembl_gene_id), ] %T>% pnrow

  # Save conversion table for future use
  saveRDS(dt_ids, file = file_dt_ids_affy_mouse430)

} else {
  dt_ids <- readRDS(file_dt_ids_affy_mouse430)
}

# Rename rows of the matrix to Ensembl identifiers using helper from jsutil / pasta
mat <- renaming_rows_in_mat_after_gene_id_conversion(mat, dt_ids, 
    species = "mouse") %T>% pdim   # 18,353 genes, 42 samples
```

## Restricting to Pasta age model genes

We next restrict the matrix to one to one mouse orthologs that are part of the Pasta age model,
and apply the rank based normalization used by the model.

```r
```{r restrict-age-model-genes, message=FALSE, warning=FALSE}
# Keep only one to one orthologs between mouse and human used by the model
mat %<>% keeping_mouse_one2one_orthologs %T>% pdim # 3,010 genes, 42 samples

# Filter to age model genes and rank normalize expression
mat %<>% filtering_age_model_genes_and_rank_norm %T>% pdim # 8,113 genes, 42 samples
```

## Extracting and cleaning phenotype data

We extract age, tissue, and strain information from the GEO phenotype data and clean the age field.

```r
```{r extract-pdata, message=FALSE, warning=FALSE}
# Phenotype data from GEOquery ExpressionSet
pdata <- pData(ES)[, c("geo_accession", "age:ch1", "tissue:ch1", "strain:ch1")] %>%
  setDT

colnames(pdata) <- c("geo_accession", "age", "tissue", "strain")

# Check that samples are aligned between expression matrix and phenotype table
all.equal(colnames(mat), pdata$geo_accession) # TRUE
```

# Age prediction with Pasta

## Adding Pasta age predictions

We now run Pasta on the transposed matrix and add predicted age scores to the phenotype data.
For this validation we only compute Pasta scores without the CT46 component.

```r
```{r age-prediction, message=FALSE, warning=FALSE}
# Adding Pasta age predictions
pdata %<>% adding_age_preds_to_pdata(t(mat)) %T>% pdim 

# Keeping only samples with non missing age
pdata <- pdata[!is.na(age)] %T>% pdim

# Converting age from character such as "6 weeks" to integer number of weeks
pdata[, age := gsub(" weeks", "", age) %>% as.integer]
```


# Results

In this section, we compute correlations between the true age and the age scores from Pasta and visualize the results.

## Correlation analysis

```{r correlations-by-group, message=FALSE, warning=FALSE}
dt_cor = pdata[, .(PCC = cor(age, PASTA)), c('tissue', 'strain')][order(-PCC)]
dt_cor[, strain := gsub('.*Apoe.*', 'Apoe', strain)]
dt_cor[, strain := gsub('.*C57B.*', 'WT', strain)]
dt_cor[, condition := paste0(tissue, ', ', strain)]
dt_cor[, condition := factor(condition, levels = dt_cor$condition %>% rev)]
dt_cor
```


## Visualization

```r
```{r plot-pcc, message=FALSE, warning=FALSE}
p1 <- ggplot(dt_cor, aes(x = PCC, y = condition)) +
  geom_point(size = 5) +
  theme_minimal(base_size = 18) +
  ggtitle("Mouse microarray (GSE40156)") +
  ylab("Tissue, strain") +
  xlab("Pearson correlation coefficient")

print(p1)
```

